foolbox/attacks/spatial_attack_transformations.py

```diff
@@ -0,0 +1,265 @@
+    theta[:, 0, 2] /= x_e.shape[2] / 2.0
+    theta[:, 1, 2] /= x_e.shape[3] / 2.0
+
+    # to pt
+    x = x_e.raw
+    theta = torch.tensor(theta, device=x.device)
+
+    assert len(x.shape) == 4
+    assert theta.shape[1:] == (2, 3)
+
+    bs, _, n_x, n_y, = x.shape
+
+    def create_meshgrid(x: torch.Tensor) -> torch.Tensor:
+        space_x = torch.linspace(-1, 1, n_x, device=x.device)
+        space_y = torch.linspace(-1, 1, n_y, device=x.device)
+        meshgrid = torch.meshgrid([space_x, space_y])  # type: ignore
+        ones = torch.ones(meshgrid[0].shape, device=x.device)
+        gridder = torch.stack([meshgrid[1], meshgrid[0], ones], dim=2)
+        grid = gridder[None, ...].repeat(bs, 1, 1, 1)[..., None]
+        return grid
+
+    meshgrid = create_meshgrid(x)
+    theta = theta[:, None, None, :, :].repeat(1, n_x, n_y, 1, 1)
+    new_coords = torch.matmul(theta, meshgrid)
+    new_coords = new_coords.squeeze_(-1)
+
+    # align_corners=True to match tf implementation
+    transformed_images = torch.nn.functional.grid_sample(  # type: ignore
+        x, new_coords, mode="bilinear", padding_mode="zeros", align_corners=True
+    )
+    return astensor(transformed_images)
+
+
+# adapted adapted from
+# https://github.com/kevinzakka/spatial-transformer-network/blob/master/stn/transformer.py
+# state @375f990 on 3 Jun 2018
+def transform_tf(
+    x_e: Tensor, translation: Tuple[float, float] = (0.0, 0.0), rotation: float = 0.0,
+) -> Any:
+    import tensorflow as tf
+
+    bs = x_e.shape[0]
+    theta = np.zeros((2, 3)).astype(np.float32)
+
+    theta[0, :] = [np.cos(rotation), -np.sin(rotation), translation[0]]
+    theta[1, :] = [np.sin(rotation), np.cos(rotation), translation[1]]
+    theta = np.tile(theta[None], (bs, 1, 1)).reshape(bs, 2, 3)
+    # convert from pixels to relative translation (bs, x, y, n_ch)
+    theta[:, 0, 2] /= x_e.shape[1] / 2.0
+    theta[:, 1, 2] /= x_e.shape[2] / 2.0
+
+    # to tf
+    theta = tf.convert_to_tensor(theta)
+    x = x_e.raw
+    """
+    Input
+    -----
+    - x: Images, should be a tensor of shape (bs, n_x, n_y, C).
+    - theta: affine transform tensor of shape (bs, 2, 3). Permits cropping,
+      translation and isotropic scaling. Initialize to identity matrix.
+      It is the output of the localization network.
+    Returns
+    -------
+    - out_fmap: transformed input feature map. Tensor of size (bs, n_x, n_y, C).
+    Notes
+    -----
+    [1]: 'Spatial Transformer Networks', Jaderberg et. al,
+         (https://arxiv.org/abs/1506.02025)
+    """
+    # grab input dimensions
+    assert theta.shape[1:] == (2, 3)
+    assert len(x.shape) == 4
+    bs = tf.shape(x)[0]
+    n_x = tf.shape(x)[1]  # height matrix
+    n_y = tf.shape(x)[2]  # width matrix
+
+    def get_pixel_value(img: Any, x: Any, y: Any) -> Any:
+        """
+        Utility function to get pixel value for coordinate
+        vectors x and y from a  4D tensor image.
+        Input
```

Comment:
Replace `Input` with `Args`. Add blank lines before `Input` and `Returns` and add a colon after `Input` and `Returns`.